cohortextractor/dsl.py

```diff
@@ -0,0 +1,185 @@
+"""This module provides classes for building a cohort using the DSL.
+
+Data lives in database tables, which are represented by PatientTable (a table with one
+row per patient) and EventTable (a table with multiple rows per patient).
+
+Through filtering, sorting, aggregating, and selecting columns, we transform instances
+of PatientTable/EventTable into instances of PatientSeries.
+
+A PatientSeries represents a mapping between patients and values, and can be assigned to
+a Cohort.  In the future, a PatientSeries will be able to be combined with another
+PatientSeries or a single value to produce a new PatientSeries.
+
+All classes except Cohort are intended to be immutable.
+
+Methods are designed so that users can only perform actions that are semantically
+meaningful.  This means that the order of operations is restricted.  In a terrible ASCII
+railway diagram:
+
+
+             +---+
+             |   V
+      filter |  PatientFrame
+             |   |   |
+             +---+   | select_column
+                     V
+                PatientSeries
+
+
+             +---+
+             |   V
+      filter |  EventFrame -----------------------+
+             |   |   |                            |
+             +---+   | sort_by                    |
+                     V                            |
+             SortedEventFrame                     |
+                     |                            |
+        +---+        | (first/last)_for_patient   | (count/exists)_for_patient
+        |   V        V                            |
+ filter |  AggregatedEventFrame                   |
+        |   |        |                            |
+        +---+        | select_column              |
+                     V                            |
+               PatientSeries <--------------------+
```

Comment:
I've just had a thought about this sequencing, which I wish I hadn't had. There are other (hypothesised) aggregations like `mean_for_patient` or `max_for_patient`, for which we would want to do `select_column` before the aggregation. Sadly.

Reply:
Don't worry, this is important to get right now, and it'll only take me a couple of hours to redraw the diagram.

`t.max_for_patient(t.value)` is equivalent to `t.sort_by(t.value).last_for_patient().select_column(t.value)`.  But this is likely to be common enough that we'd want to provide some sugar.

However, there's no equivalent of `mean_for_patient` or any other SQL aggregate functions like `SUM`.

Related, the DSL doesn't currently support passing columns to `count_for_patient` and `exists_for_patient`, so to find whether a patient has any records with a non-null value, you have to do something like `t.filter("value", include_null=True).exists_for_patient()`.  We may want to provide sugar here too.

So my proposal is to tweak/add methods to `EventFrame`:

* We tweak `(count|exists)_for_patient` so they now take an optional column (defaulting to `"patient_id"`).
* We add `(min|max|mean|sum)_for_patient`, which take a required column.
* We add `(first|last)_for_patient`, which take a required column, and which are sugar as described able.  We don't remove `SortedEventFrame.(first|last)_for_column`, since we need to support eg finding the code of the latest event.

Reply:
I like this proposal. One possible counter is to do something like `aggregate_for_patient(agg.first)` but I think I prefer your suggestion. We're already going a bit domain-specific by talking about patients, so let's keep being domain-specific with our aggregation options....?

Reply:
Do we need the `(first|last)_for_patient` syntactic sugar for min/max?  Those are sqlalchemy functions, can't they all just be passed to `qm_table.aggregate()`?

Is `SortedEventFrame.(first|last)_for_column` a typo (of `for_patient`)?

Reply:
> Do we need the `(first|last)_for_patient` syntactic sugar for min/max?

Good point!  I hadn't realised that `min_for_patient` would be equivalent to `first_for_patient`.  We should probably make `min`/`max` the defaults, and then offer `first`/`last` as aliases.

> Is SortedEventFrame.(first|last)_for_column a typo (of for_patient)?

Yes, fixed.